File: CS/Q234787/MyDateEdit.cs
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;
using System.ComponentModel;
using DevExpress.XtraEditors.Popup;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.Calendar;
using System;

namespace DXSample {
    public class MyDateEdit : DateEdit {
        static MyDateEdit() { RepositoryItemMyDateEdit.RegisterMyDateEdit(); }
        public MyDateEdit() : base() { }

        public override string EditorTypeName { get { return RepositoryItemMyDateEdit.MyDateEditName; } }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMyDateEdit Properties { get { return (RepositoryItemMyDateEdit)base.Properties; } }

        protected override PopupBaseForm CreatePopupForm() {
            if (!Properties.IsVistaDisplayModeInternal)
                return new MyPopupDateEditForm(this);
            return base.CreatePopupForm();
        }
    }

    [UserRepositoryItem("RegisterMyDateEdit")]
    public class RepositoryItemMyDateEdit : RepositoryItemDateEdit {
        static RepositoryItemMyDateEdit() { RegisterMyDateEdit(); }
        public RepositoryItemMyDateEdit() : base() { }

        internal const string MyDateEditName = "MyDateEdit";

        public override string EditorTypeName { get { return MyDateEditName; } }

        public static void RegisterMyDateEdit() {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(MyDateEditName, typeof(MyDateEdit),
                typeof(RepositoryItemMyDateEdit), typeof(DateEditViewInfo), new ButtonEditPainter(), true));
        }

        internal bool IsVistaDisplayModeInternal { get { return IsVistaDisplayMode(); } }
    }

    public class MyPopupDateEditForm : PopupDateEditForm {
        public MyPopupDateEditForm(DateEdit ownerEdit) : base(ownerEdit) { }

        protected override DateEditCalendar CreateCalendar() {
            return new MyDateEditCalendar(OwnerEdit.Properties, OwnerEdit.EditValue);
        }
    }

    public class MyDateEditCalendar : DateEditCalendar {
        public MyDateEditCalendar(RepositoryItemDateEdit item, object editDate) : base(item, editDate) { }

        protected override DateEditInfoArgs CreateInfoArgs() {
            return new MyDateEditInfoArgs(this);
        }
    }

    public class MyDateEditInfoArgs : DateEditInfoArgs {
        public MyDateEditInfoArgs(MyDateEditCalendar calendar) : base(calendar) { }

        protected override int GetWeekNumber(DateTime date) {
            DateTime baseDate = date.Month < 10 ? new DateTime(date.Year - 1, 10, 1) : new DateTime(date.Year, 10, 1);
            return (int)Math.Floor((double)(new DateTime(date.Ticks - baseDate.Ticks).DayOfYear / 7));
        }
    }
}

File: VB/Q234787/MyDateEdit.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing
Imports System.ComponentModel
Imports DevExpress.XtraEditors.Popup
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.Calendar
Imports System

Namespace DXSample
	Public Class MyDateEdit
		Inherits DateEdit
		Shared Sub New()
			RepositoryItemMyDateEdit.RegisterMyDateEdit()
		End Sub
		Public Sub New()
			MyBase.New()
		End Sub

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemMyDateEdit.MyDateEditName
			End Get
		End Property

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As RepositoryItemMyDateEdit
			Get
				Return CType(MyBase.Properties, RepositoryItemMyDateEdit)
			End Get
		End Property

		Protected Overrides Function CreatePopupForm() As PopupBaseForm
			If (Not Properties.IsVistaDisplayModeInternal) Then
				Return New MyPopupDateEditForm(Me)
			End If
			Return MyBase.CreatePopupForm()
		End Function
	End Class

	<UserRepositoryItem("RegisterMyDateEdit")> _
	Public Class RepositoryItemMyDateEdit
		Inherits RepositoryItemDateEdit
		Shared Sub New()
			RegisterMyDateEdit()
		End Sub
		Public Sub New()
			MyBase.New()
		End Sub

		Friend Const MyDateEditName As String = "MyDateEdit"

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return MyDateEditName
			End Get
		End Property

		Public Shared Sub RegisterMyDateEdit()
			EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(MyDateEditName, GetType(MyDateEdit), GetType(RepositoryItemMyDateEdit), GetType(DateEditViewInfo), New ButtonEditPainter(), True))
		End Sub

		Friend ReadOnly Property IsVistaDisplayModeInternal() As Boolean
			Get
				Return IsVistaDisplayMode()
			End Get
		End Property
	End Class

	Public Class MyPopupDateEditForm
		Inherits PopupDateEditForm
		Public Sub New(ByVal ownerEdit As DateEdit)
			MyBase.New(ownerEdit)
		End Sub

		Protected Overrides Function CreateCalendar() As DateEditCalendar
			Return New MyDateEditCalendar(OwnerEdit.Properties, OwnerEdit.EditValue)
		End Function
	End Class

	Public Class MyDateEditCalendar
		Inherits DateEditCalendar
		Public Sub New(ByVal item As RepositoryItemDateEdit, ByVal editDate As Object)
			MyBase.New(item, editDate)
		End Sub

		Protected Overrides Function CreateInfoArgs() As DateEditInfoArgs
			Return New MyDateEditInfoArgs(Me)
		End Function
	End Class

	Public Class MyDateEditInfoArgs
		Inherits DateEditInfoArgs
		Public Sub New(ByVal calendar As MyDateEditCalendar)
			MyBase.New(calendar)
		End Sub

		Protected Overrides Function GetWeekNumber(ByVal [date] As DateTime) As Integer
			Dim baseDate As DateTime
			If [date].Month < 10 Then
				baseDate = New DateTime([date].Year - 1, 10, 1)
			Else
				baseDate = New DateTime([date].Year, 10, 1)
			End If
			Return CInt(Fix(Math.Floor(CDbl(New DateTime([date].Ticks - baseDate.Ticks).DayOfYear / 7))))
		End Function
	End Class
End Namespace